From 585179d8997b1fb8695b443f8eb94447ec3e9677 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:35 -0400
Subject: [virt] kvm/nvmx: Unconditionally uninit the MMU on nested vmexit

Message-id: <1406040016-3289-121-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86199
O-Subject: [RHEL7 PATCH v2 120/221] KVM: nVMX: Unconditionally uninit the MMU on nested vmexit
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@web.de>

Three reasons for doing this: 1. arch.walk_mmu points to arch.mmu anyway
in case nested EPT wasn't in use. 2. this aligns VMX with SVM. But 3. is
most important: nested_cpu_has_ept(vmcs12) queries the VMCS page, and if
one guest VCPU manipulates the page of another VCPU in L2, we may be
fooled to skip over the nested_ept_uninit_mmu_context, leaving mmu in
nested state. That can crash the host later on if nested_ept_get_cr3 is
invoked while L1 already left vmxon and nested.current_vmcs12 became
NULL therefore.

Cc: stable@kernel.org
Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit 29bf08f12b2fd72b882da0d85b7385e4a438a297)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 4e467d9..e095276 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -8353,8 +8353,7 @@ static void load_vmcs12_host_state(struct kvm_vcpu *vcpu,
  vcpu->arch.cr4_guest_owned_bits = ~vmcs_readl(CR4_GUEST_HOST_MASK);
  kvm_set_cr4(vcpu, vmcs12->host_cr4);
 
- if (nested_cpu_has_ept(vmcs12))
-  nested_ept_uninit_mmu_context(vcpu);
+ nested_ept_uninit_mmu_context(vcpu);
 
  kvm_set_cr3(vcpu, vmcs12->host_cr3);
  kvm_mmu_reset_context(vcpu);
-- 
1.7.1